### ISSUE TITLE
scanner-v4-matcher exceeds memory limit and is killed

### ISSUE THREAD

**Issue by ghoff** [stackrox/stackrox#11553]
I've enabled scanner v4 in in stackrox 4.4.x.  In at least 4.4.2 and 4.4.3 I've seen the matcher fail when consuming more than the  5Gi memory limit.  Editing the deployment and increasing the memory limit to 6 allows it complete indexing.
This is observed in a test environment running on k3s 1.29.5+k3s1.  As this is test, k3s is not running constantly.  I don't know if this is a contributing factor.

These are the helm flags I used for scanner v4

--set scannerV4.indexer.autoscaling.disable=true --set scannerV4.indexer.replicas=1 \
--set scannerV4.matcher.autoscaling.disable=true --set scannerV4.matcher.replicas=1 \
--set scannerV4.disable=false

**Comment by ghoff** [comment]
Example of it exceeding 5GB of ram
```
$ kubectl top pod -n stackrox -l app=scanner-v4-matcher --sort-by=cpu
NAME                                 CPU(cores)   MEMORY(bytes)   
scanner-v4-matcher-5c5c55fc6-j9bv6   1965m        5501Mi 
```

This is a series of items in the pod logs at the time

{"level":"info","host":"scanner-v4-matcher-5c5c55fc6-j9bv6","component":"libvuln/OfflineImporter","updater":"RHEL7-openstack-8-including-unpatched","ref":"4740f873-0037-4695-b9f4-dffe0deeaad9","vuln_count":6708,"enrichment_count":0,"time":"2024-06-14T15:22:45Z","message":"update imported"}
{"level":"info","host":"scanner-v4-matcher-5c5c55fc6-j9bv6","component":"libvuln/OfflineImporter","updater":"RHEL9-openshift-4.14","ref":"64d9e8e2-59e6-413b-a0e9-f0dfcdced3c5","vuln_count":432,"enrichment_count":0,"time":"2024-06-14T15:22:45Z","message":"update imported"}
{"level":"info","host":"scanner-v4-matcher-5c5c55fc6-j9bv6","component":"libvuln/OfflineImporter","updater":"RHEL7-jboss-ws-5","ref":"0d1461ca-25af-4d2f-b4a1-1c253ee24e15","vuln_count":4656,"enrichment_count":0,"time":"2024-06-14T15:22:45Z","message":"update imported"}
{"level":"info","host":"scanner-v4-matcher-5c5c55fc6-j9bv6","component":"libvuln/OfflineImporter","updater":"RHEL6-rhel-6.1-eus","ref":"049b18de-5822-4898-b993-b6e18cc3807d","vuln_count":2830,"enrichment_count":0,"time":"2024-06-14T15:22:46Z","message":"update imported"}
{"level":"info","host":"scanner-v4-matcher-5c5c55fc6-j9bv6","component":"libvuln/OfflineImporter","updater":"RHEL9-openstack-17.1","ref":"0f81cc9c-979b-4c56-94e0-2397de472b4a","vuln_count":56,"enrichment_count":0,"time":"2024-06-14T15:22:46Z","message":"update imported"}

**Comment by RTann** [comment]
This is a known issue with the Matcher, and it one of the main reasons why it was originally deployed in Tech Preview. There has been work to resolve this, and we have found the memory usage is now in the order of 100s of MB instead of GB these days. However, as mentioned, this is for the Matcher.

Can you tell us more about the Indexer? You sent over the logs for the Matcher, so I'm wondering if you meant Matcher instead of Indexer. If not, then I'm wondering if you can tell us more about the images you scanned (size, perhaps number of images)

**Comment by ghoff** [comment]
I'm sorry, I very much meant the matcher.  I'll update the issue to correct that.  This is a clean install of k3s which has these images in kube-system
docker.io/rancher/mirrored-coredns-coredns:1.10.1
docker.io/rancher/mirrored-library-traefik:2.10.7
docker.io/rancher/local-path-provisioner:v0.0.26
docker.io/rancher/klipper-lb:v0.4.7
docker.io/rancher/mirrored-metrics-server:v0.7.0

it has the images from stackrox 4.4.2, now inactive, and the images used by 4.4.3

**Comment by RTann** [comment]
@ghoff has this been resolved? The Scanner V4 matcher's current request/limit is 500MB/2GB https://github.com/stackrox/stackrox/blob/4.5.5/image/templates/helm/stackrox-central/internal/defaults.yaml.htpl#L262

**Comment by porridge** [comment]
@ghoff can we consider this issue solved, then? Or is there something else left to do here?

**Comment by ghoff** [comment]
Yes, I consider it solved.  Thank you for the time put in on resolving this.
